rust/nix-expr/src/eval_state.rs

```diff
@@ -106,6 +117,102 @@ impl EvalState {
         let r = unsafe { raw::nix_get_type(self.context.ptr(), value.raw_ptr()) };
         Ok(ValueType::from_raw(r))
     }
+    pub fn require_int(&self, v: &Value) -> Result<Int> {
+        let t = self.value_type(v).unwrap();
+        if t != ValueType::Int {
+            bail!("expected an int, but got a {:?}", t);
+        }
+        let i = unsafe { raw::nix_get_int(self.context.ptr(), v.raw_ptr()) };
+        Ok(i)
+    }
+    /// Evaluate, and require that the value is an attrset.
+    /// Returns a list of the keys in the attrset.
+    pub fn require_attrs_names(&self, v: &Value) -> Result<Vec<String>> {
+        let t = self.value_type(v)?;
+        if t != ValueType::AttrSet {
+            bail!("expected an attrset, but got a {:?}", t);
+        }
+        let n = unsafe { raw::nix_get_attrs_size(self.context.ptr(), v.raw_ptr()) as usize };
+        self.context.check_err()?;
+        let mut attrs = Vec::with_capacity(n);
+        unsafe {
+            for i in 0..n {
+                let cstr_ptr: *const i8 = raw::nix_get_attr_name_byidx(
+                    self.context.ptr(),
+                    v.raw_ptr(),
+                    self.raw_ptr(),
+                    i as c_uint,
+                );
+                self.context.check_err()?;
+                let cstr = std::ffi::CStr::from_ptr(cstr_ptr);
+                let s = cstr.to_str().map_err(|e| {
+                    anyhow::format_err!("Nix attrset key is not valid UTF-8: {}", e)
+                })?;
+                attrs.insert(i, s.to_owned());
+            }
+        }
+        Ok(attrs)
+    }
+
+    pub fn require_attrs_select(&self, v: &Value, attr_name: &str) -> Result<Value> {
+        let r = self.require_attrs_select_opt(v, attr_name)?;
+        match r {
+            Some(v) => Ok(v),
+            None => self.context.check_err().and_then(|_| {
+                // should be unreachable
+                bail!("attribute not found: {}", attr_name)
+            }),
+        }
+    }
+
+    /// Evaluate, require that the value is an attrset, and select an attribute by name.
+    pub fn require_attrs_select_opt(&self, v: &Value, attr_name: &str) -> Result<Option<Value>> {
+        let t = self.value_type(v)?;
+        if t != ValueType::AttrSet {
+            bail!("expected an attrset, but got a {:?}", t);
+        }
+        let attr_name = CString::new(attr_name)
+            .with_context(|| "require_attrs_select_opt: attrName contains null byte")?;
+        // c_void should be Value; why was void generated?
```

Comment:
because of `typedef void Value` in [`nix_api_expr.h`](https://github.com/roberth/nix/blob/a99aa8b6f3dd05f12c7613de1f08d1067836d267/src/libexpr-c/nix_api_expr.h#L56). your question, https://github.com/NixOS/nix/pull/8699#discussion_r1264698086, why it is definied like this was never answered, was it?

Reply:
I *think* it could just be improved, and I've opened an issue, but I'll throw it in the [Nix Bindings channel](https://matrix.to/#/#nix-bindings:nixos.org) as well to be sure.
- https://github.com/NixOS/nix/issues/10561